sv-run.8
.\"
.\" CopyLeft (c) 2016 tokiclover <tokiclover@gmail.com>
.\"
.\" Distributed under the terms of the 2-clause BSD License as
.\" stated in the COPYING file that comes with the source files
.\"
.pc
.TH SV-RUN 8 "2018-08-20" "0.14.0" "System Manager's Manual"
.SH NAME
sv-run \- service manager utility
.SH DESCRIPTION
.B sv-run
is a utility to manage runscript and supervision services.
This utility is used in conjunction with a supervisor like
.B runit
,
.B s6
or
.RB \| daemontools [ \-encore ] \|
to get a fully supervised init-system and service-manager.
.SH SYNOPSIS
.B sv-run
.RB [\| OPTIONS \| ]
.RB \| SERVICE \|
.RB \| COMMAND \|
.RB [\| ARGUMENTS \|]

.B sv-run
.RB [\| \-h|\-\-help \|]
.RB [\| \-v|\-\-version \|]

.SH SERVICE
Service name can be replaced by
.B scan
to update the dependency cache of services passed as arguments
like \fBsv-run scan \fIspawn-fcgi.{lighttpd,nginx}\fR.
When no service name is specified in the command line, the whole service
dependency cache is updated.

Service name can be replaced by the special use case
.B init
to manage a whole bundle of services instead of a single service; for example,
`sv-run --default init' to bring the system to default runlevel.
This usage is now superseded by
.B sv-rc
utility, see
.BR sv-rc (8)
manual page; and is thus discouraged.

.B NOTE:
If
.B service
is not installed in the default path, either add
.B @LIBDIR@/bin:@LIBDIR@/sbin
to
.B PATH
environment variable; or else, copy the symlink to
.I @EXEC_PREFIX@/sbin
for example to get the same behaviour; or else use
.B sv-run
instead because
.B service
is a just limited interfaces to manage service.

Do note that it's rather prefered and recommended to use
.B sv-run
command to manage services rather than using the service executable. And this
is more true for supervision services when dealing with dependencies and
virtual services. Starting a supervision service shall be done with
.B sv-run
binary instead of using
.B @SYSCONFDIR@/sv/SERVICE/run
.RB \| COMMAND \|
.RB [\| ARGUMENTS \|]

.SH OPTIONS
.TP
.RB (\| \-\-sysinit | \-\-sysboot | \-\-default | \-\-shutdown | \-\-reboot | \-\-single | \-\-nonetwork \|)
Select a particular init stage or runlevel (long options)
.TP
.RB (\| \-0 | \-1 | \-2 | \-3 | \-4 | \-5 | \-6 \|) | (\| \-p | \-s | \-N | \-l | \-S | \-b | \-r \|)
Select a particular init stage or runlevel (short options)
.TP
.RB \| \-D | \-\-nodeps \|
Disable dependency tracking when starting/stopping services
.TP
.RB \| \-d | \-\-debug \|
Enable debugging mode for services (debugging output to stderr)
.TP
.RB \| \-x | \-\-trace \|
Enable shell trace for services (shell tracing)
.TP
.RB \| \-q | \-\-quiet \|
Suppress verbose message
.TP
.RB \| \-v | \-\-version \|
Print the version message
.TP
.RB \| \-h | \-\-help \|
Print the help message
.SH COMMANDS
Standard service commands along with extra service commands are supported.
.TP
.B start
Start a service if down or stopped (supervision service wait up to 10 seconds
by default to start a service.)
.TP
.B stop
Stop a service if up or started (supervision service wait up to 10 seconds by
default to stop a service.)
.TP
.B restart
Restart a service after stopping it if started (runscript service can define
a particular shell function for this purpose.)
.TP
.B remove
Remove a supervised service from the running directory.
This command has no effect for runscript services.
.TP
.B status
Report the status of service (if started or stopped.)
.TP
.B zap
Remove the temporary files of a service to be able to start a fresh environment,
or restart a service if already started.
.TP
.B add
Add a service to an init-stage or runlevel (RUNLEVEL argument is required.)
.TP
.B del
Remove a service from an init-stage or runlevel (stage level argument is required.)
.TP
.B cgroup_remove_service
Remove the Control Group
hierarchy after stopping a service. This command has no effect for runscript
services if not used to manage a daemon.
.B CGROUP_CLEANUP=ENABLE
environmenet variable can be enabled for automatic clean up.
.TP
.B new
add a new supervised service from a template. This command should be run with
`\fI@LIBDIR@/sv/bin/sv-config \fR[\fI\-\-log\fR] \fBSERVICE \fInew\fR'.
.TP
.B EXTRA_COMMANDS
Extra comands defined as shell functions and added to
.B SVC_COMMANDS
(command that can be executed with any service status) or
.B SVC_STARTED_COMMANDS
(for any command that shoould be executed in started service status) or
.B SVC_STOPPED_COMMANDS
(for command that should be executed in stopped service status)
environment variable. Any defined command can be used like a standard service command.
.TP
.B desc
.RB [\| COMMAND \|]

Print service description to standard output.
.B COMMAND
argument can be passed to print the description of the named command instead of the
description of the service, `sv-run udev desc reload' for example.
.TP
.B scan
Update the dependency cache of a service. \fIsv-run scan \fBSERVICES\fR can be used
for a list of service instead.
.TP
.B down
Send down command to the supervisor (supervision only)
.TP
.B up
Send up command to the supervisor (supervision only)
.SH CONFIGURATION
See
.I @SYSCONFDIR@/sv.conf
for the global configuration file; or
.I @SYSCONFDIR@/sv.conf.d/\fBSERVICE\fR
for runscript service configuration file; or
.I @SYSCONFDIR@/sv/\fBSERVICE\fI/OPTIONS\fR[\fB.INSTANCE\fR]
for supervision service configuration file; or else,
.I @SYSCONFDIR@/sv.conf.local.d/\fBSERVICE\fR
for the local configuration file for any service type.
See
.BR supervsion (5)
for settings related to
.B sv-run
service management.
.SH FILES
.SS @SYSCONFDIR@
.TP
.I @SYSCONFDIR@/sv.conf
Supervision global configuration file
.TP
.B @SYSCONFDIR@/sv/SERVICE
Service file or directory for supervision
.TP
.B @SYSCONFDIR@/sv.conf.d/SERVICE
Service configuration file (RUNSCRIPT SERVICE)
.TP
.RB \| @SYSCONFDIR@/sv/SERVICE/OPTIONS [ .INSTANCE ] \|
Service configuration file (SUPERVISION SERVICE)
.TP
.B @SYSCONFDIR@/sv.conf.local.d/SERVICE
Local service configuration file (for any service type)
.SS @LIBDIR@/sv
.TP
.I @LIBDIR@/sv/sh/SV-CONFIG
Specific supervision environment configuration file
.TP
.B @LIBDIR@/sv/sh/run
Default
.B @SYSCONFDIR@/sv/\c
.RB \| SERVICE/ { log/ }{ finish , run } \|
command utility
.TP
.B @LIBDIR@/sv/sh/cgroup-functions
CGroup Functions/helper library
.TP
.B @LIBDIR@/sv/sh/functions
Generic Functions/Helpers library
.TP
.B @LIBDIR@/sv/sh/sv-run.sh
Runscript shell script utility
.TP
.B @LIBDIR@/sv/sh/runscript-functions
Runscript Functions/helpers library
.TP
.B @LIBDIR@/sv/sh/supervision-functions
Supervision Functions/helpers library

.SH "RETURN VALUE"
.TP
.B 0
success (service already started or a service commmand was successfully executed)
.TP
.B 1
miscellaneous operational failures; or
a signal was sent to abort or terminate the program; or
a syscall failed;
.TP
.B 2
failure to find the requested service
.TP
.B 3
invalid arguments (trying to stop a non started service; add or remove a service
to a runlevel without specifying the runlevel...)
.TP
.B 4
failure to lock a service lockfile for service command; or timed out when trying to
lock the lockfile.
.TP
.B 5
service keywords does not permit to {start,stop} the service
.TP
.B 7
for
.I status
service when a service is stopped
.TP
.B 8
for
.I status
service command when a service is active and down (supervision only)
.TP
.B 9
for
.I status
service command when a service command failed
.TP
.B 10
for
.I status
service command when a service is waiting for the lockfile
.SH "APPLICATION USAGE"
Some services are added to the service directory? Try starting them then.
It seems to work? Good, try to add them in an init runlevel with
`\fIsv-run \fB\-\-RUNLEVEL SERVICE \fIadd\fR'; and then regenerate the dependency
scheduling priority file by running `\fIsv-rcorder \fB\-\-RUNLEVEL \fIscan\fR'
and then checkout the output to see if the service is inserted
in the right place depending on the dependency definitions. Or better,
check out the ordering beforehand with `\fIsv-rcorder \fB\-\-RUNLEVEL\fR'.
It looks weird or the order changed dramatically?
Try again\-\-update the dependency definitions\-\-and then issue
`\fIsv-rcorder \fB\-\-RUNLEVEL \fIscan\fR' to update the dependency cache and
the scheduling priority ordering. Else, it can be a case of cyclical
dependency definition. Major dependency redefinition is necessary to fix this
issue which will prevent
.BR sv-rc (8)
to schedule and order the start/stop of services.
.SH "FUTURE DIRECTIONS"
None.
.SH "SEE ALSO"
.BR supervision (5),
.BR sv-init (8),
.BR sv-rc (8),
.BR sv-rcorder (8),
.BR sv-shutdown (8),
.SH AUTHORS
tokiclover <tokiclover@supervision.project>
.\"
.\" vim:fenc=utf-8:ft=groff:ci:pi:sts=2:sw=2:ts=2:expandtab:
.\"


ChangeLog
Header: supervision/ChangeLog, 2018/08/16

0.14.7:

Memory alignement fix (fix structures padding);
Fix miultiple logfile open (important fix);
Added a workaround to LVM pvscan(8) stall with udev;

0.14.6:

Fix redundant dependency start up;
Fix logging to log file (important fix);
Fix zramfs infinite loop (0.14.5);

0.14.5:

Renamed loopback service to loop-back-device;

Tiny tixes to try to make LVM works out with udev.
Synchronization issue may happen because udev is not ready to initialize
devices; and then LVM will block the boot process.
Remain hitting the ^C (CTRL-C) to halt lvm.

0.14.4:

Fixed a nasty structure initilization bug (sv-run);

0.14.3:

Bug fixes: most important is svc_run() pid fix in sv-run(8);
Renamed `sv.log' logfile to `sv-rc.log';

0.14.2:

Merged remaining commits from master branch into this stable;
Some commits were missing from previous v0.14.1. **Apology**

0.14.1:

`sv-rc(8)' bug fix (environment variables were not being compared properly),
so SV_TIMEOUT was being defined with the value of SV_TIMEOUT_*, nasty bug;

0.14.0:

Removed `/lib/sv/opt' and merged the directory to `/lib/sv/sh';
Renamed `rs(8)' command to `sv-run(8)';
Renamed `sv-stage(8)' command to `sv-rc(8)'; along with the
environment variable `SV_STAGE' to `SV_INITLEVEL';
Renamed `OS_NAME' to `SV_UNAME' and added `SV_UNAME_RELEASE'
(defined in `${SV_TMPDIR}/env');
Renamed `runscript' shell script to `sv-run.sh';
Renamed `init-stage' shell script to `sv-init.sh';
Renamed `depgen' shell script to `sv-deps.sh';
Renamed `cmd' (supervision shell script) to `run';
Renamed `getty' service instances rightly with a `.' (dot)
instead of a hyphen;

*IMPORTANT NOTE**: if getty service are started, the sysadmin should add
the old symbolc links with
`for s in /etc/sv/getty-tty{2,3,4,5,6,S0,S1}; do ln -fs getty.tty1 $d; done'
before shutting down the service or system!


Manpages was cleaned and updated to v0.14.0;
See `supervision(5)' manual page for more information about the changes;
And also, `sv-run(8)' and `sv-rc(8)' manual pages;

A few bug fixes were ported to v0.13.1 (not all the bugs which require more
work; A nasty thread race bug was fixed lastly which was causing `sv-rc' to hang
waiting a SIGCHLD from a dead child;

`sv-config' was refactorized and coded in C; and now has a -u|--update command
line switch; a `sv-config --update' is now required when updating an old
install to handle the API/ABI changes from previous v0.13.0 and the changes
between v0.12.0 and v0.13.0;

The old `rs stage [COMMAND]' is now replaced with `sv-run init [COMMAND]'
which is now discouraged from general usage;
`sv-rc(8)' takes a runlevel argument like *standard* `rc(8)' utility;
The old `sv-rc --RUNLEVEL' (from previous release) is also supported;

New debugging code was added and can be enabled at runtime with `SV_DEBUG' or
`SVC_DEBUG' envronment variable;
An extra `SV_DEBUG_INFO' environement variable is available to service
information level message;
For more verbose debugging code, configure with `--enable-debug';
For logging debugging output: enable `SV_SYSLOG' and `SV_SYSLOG_INFO';
and then enable `SV_LOGGER' to save logging and debugging output to
`/var/log/sv.log';

ADDIDION:
SUPERVISION SERVICE:
RUNSCRIPT SERVICE: DEFAULT, SYSBOOT, local, qemu-vlan.vnet{0,4};

0.13.1:

Stabilisation release: there are fixes alll over the place, a few are
important and many are minors.

Important fixes:
- remove virtual service from dependency tree;
- remove failure and environment service state file before `start' command;
- `zap' command remove service pidfile;
- set up `__cmd__=finish' for supervision;
- added a (shell) `trap' for service state files;
- `fstabinfo' output any request;
- use standard GNU style `__attribute__';
- remove faulty `SV_PREFIX' from `sv-stage';
- and minor fixes, see `git log' for more;

0.13.0:

WARING: SHORT COMMAND LINE OPTION FOR RUNLEVELS CHANGED DRAMATICALLY FROM v.12.x

Improve signal handling so SIG{INT,QUIT,TERM} are not blocked anymore;
Improve parallelization of service (to avoid CHILD_MAX hard limit) and improve
scalabilty to thousends of service in parallel (POSIX threads);

Added support for keywords and timeout per service; so a NOHANG option is now
available to kill hangged service (see sv.conf for more iformation);
Keywords can be used to disable services per subsystem (like docker...);

Added support for containrization solutions or subsystems via keywords
usage (see KEYWORDS subsection in supervision(5) and sv.conf for more information)
for docker, LXC, jail, systemd-nspawn, prefix, supervision, UML, VServer and XEN.
Either the subsystem will be auto detected or use sv.conf to set a particular
subsystem with `SV_SYSTEM="${SUBSYSTEM}"' configuration variable...
`SV_SYSTEM="supervision"' for daemon supervision only;
`SV_SYSTEM="prefix"' for an isolated chrooted environment;
`SV_SYSTEM="docker"' for docker containers et al.
Services that have the subsystem keyword will not be started in that particular
subsystem environment.

To have the supervisor `({damontools[-encore],runit,s6})' executed as __PID 1__...
Just setup the container or subsystem; once done, use something like
the following for docker: `docker run [OPTIONS] --env container=docker --tmpfs /run
IMAGE /lib/sv/sh/init-stage --default'; and voila! the supervisor will be executed
as `PID 1' and another process will handle service management to setup the container.

Added abilty to abort an init-stage (used to abort and bring single runlevel
when checkfs service fails to clean up filesystems;

Added new manual to document new sv-stage(8) binary and refactorized rs(8);
Added a status command for init-stage/runlevel to retrieve service status:
e.g. `sv-stage --default status' to get service status of default runlevel;

Use POSIX threads to improve performance with parallel start/stop instead of
using only process based technics (child process/fork(2));


lib/sh/sv-run.sh
#!/bin/sh
#
# $Id:  @(#) sv-run.sh    0.8 2018/07/22 21:09:26                     Exp $
# $C$:  Copyright (c) 2015-2018 tokiclover <tokiclover@gmail.com>     Exp $
# $L$:  2-clause/new/simplified BSD License                           Exp $
#

readonly __arg__="$*"
#
# ZSH compatibility
#
if [ -n "${ZSH_VERSION}" ]; then
	emulate sh
	NULLCMD=:
	setopt NO_GLOB_SUBST SH_WORD_SPLIT
	disable -r end
fi

help_message() {
	cat <<-EOH
 usage: ${0##*/} [OPTIONS] SERVICE COMMAND [ARGUMENTS]
   COMMAND: add|del|desc|remove|reload|restart|start|stop|status|zap|cgroup_remove_service
   OPTIONS: [OPTIONS] SERVICE COMMAND [ARGUMENTS]
     -D, --nodeps     Disable dependencies1
     -d, --debug      Enable debug mode
     -x, --trace      Enable shell trace
     -r, --rs         Select runscript service type
     -s, --sv         Select supervision service type
     -h, --help       Print help and exit
EOH
${1+exit ${1}}
}

while true; do
	case "${1}" in
		(-D|--node*) SVC_DEPS=0 ;;
		(-p|--deps*) SVC_DEPS=1 ;;
		(-r|--rs) SV_TYPE="rs" ;;
		(-s|--sv) SV_TYPE="sv" ;;
		(-x|--trace) set -x ;;
		(-d|--debug) SVC_DEBUG=true;;
		(-h|--help) help_message 0 ;;
		(*) break ;;
	esac
	shift
done
if ! [ ${#} -ge 2 ]; then
	echo "${0##*/}: error: Insufficient number of arguments" >&2
	exit 1
fi

:	${SV_LIBDIR:=${0%/sh/*}}
:	${SVC_NAME:=${1##*/}}
name="${SVC_NAME}"
readonly __cmd__="${2}"
readonly __cmd_args__="${@#*$2}"
readonly __svc__="${1}"
readonly __av0__="${0}"
if [ -d "${__svc__}" ]; then
:	${SV_TYPE:=sv}
else
:	${SV_TYPE:=rs}
fi
case "${1}" in
	(/etc/sv/*)
:	${SV_SVCDIR:=${1%/*}}
	;;
esac
shift
readonly SVC_NAME SV_TYPE

if ! . ${SV_LIBDIR}/sh/runscript-functions; then
	echo "${0##*/}: error: Required file not found \`${SV_LIBDIR}/sh/runscript-functions'" >&2
	rm -f ${SV_TMPDIR}/wait/${SVC_NAME}
	exit 1
fi

svc_cleanup()
{
	local ret="$?" last="$_"
	rm -f ${SV_TMPDIR}/wait/${SVC_NAME} ${SV_TMPDIR}/envs/${SVC_NAME}
	_="${last}"
	return ${ret}
}
trap svc_cleanup INT TERM QUIT
SVC_DEBUG "${__svc__}:pid=$$=( ${@} )"

#
# Handle service command
#
if   [ "${SV_TYPE}" = "rs" ]; then
	rs_cmd "${@}"
elif [ "${SV_TYPE}" = "sv" ]; then
	#
	# Save environment file to handle false start failures
	#
	if [ "${__cmd__}" = "start" -o "${__cmd__}" = "stop" ] || \
	   [ "${__cmd__}" = "up"    -o "${__cmd__}" = "down" ]; then
	while read line; do
		echo "${line}"
	done <${SV_TMPDIR}/environ >${SV_TMPDIR}/envs/${SVC_NAME}
	ENV_SET SVC_DEPS
	else
	rm -f ${SV_TMPDIR}/envs/${SVC_NAME}
	fi

	svc_cmd "${@}"
else
	error "nothing to do -- invalid usage"
	exit 1
fi
